appstudio-controller/controllers/appstudio.redhat.com/sandboxprovisioner_controller.go

```diff
@@ -0,0 +1,211 @@
+		if err != nil {
+			log.Error(err, "failed to create a new SpaceRequest for the DeploymentTargetClaim")
+			return ctrl.Result{}, err
+		}
+	}
+
+	log.Info("A SpaceRequest for the DeploymentTargetClaim exists", "SpaceRequest.Name", spaceRequest.Name, "Namespace", spaceRequest.Namespace)
+
+	return ctrl.Result{}, nil
+}
+
+func missingDTCLSErrWrap(dtcName, dtclsName string) func(string) error {
+	return func(msg string) error {
+		return fmt.Errorf("DeploymentTargetClaim %s does not have a matching DeploymentTargetClass %s: %s", dtcName, dtclsName, msg)
+	}
+}
+
+// findMatchingDTClassForDTC tries to find a DTCLS that matches the given DTC in a namespace.
+func findMatchingDTClassForDTC(ctx context.Context, k8sClient client.Client, dtc applicationv1alpha1.DeploymentTargetClaim) (*applicationv1alpha1.DeploymentTargetClass, error) {
+	dtclsList := applicationv1alpha1.DeploymentTargetClassList{}
+	if err := k8sClient.List(ctx, &dtclsList); err != nil {
+		return nil, err
+	}
+
+	var dtcls *applicationv1alpha1.DeploymentTargetClass
+	for i, d := range dtclsList.Items {
+		if d.Name == string(dtc.Spec.DeploymentTargetClassName) {
+			dtcls = &dtclsList.Items[i]
+			break
+		}
+	}
+
+	return dtcls, nil
+}
+
+// findMatchingSpaceRequestForDTC tries to find a SpaceRequest that matches the given DTC in a namespace.
+// The function will return only the SpaceRequest that matches the expected environment Tier name
+func findMatchingSpaceRequestForDTC(ctx context.Context, k8sClient client.Client, dtc *applicationv1alpha1.DeploymentTargetClaim) (*codereadytoolchainv1alpha1.SpaceRequest, error) {
+	spaceRequestList := codereadytoolchainv1alpha1.SpaceRequestList{}
+
+	opts := []client.ListOption{
+		client.InNamespace(dtc.Namespace),
+		client.MatchingLabels{
+			deploymentTargetClaimLabel: dtc.Name,
+		},
+	}
+
+	if err := k8sClient.List(ctx, &spaceRequestList, opts...); err != nil {
+		return nil, err
+	}
+
+	if len(spaceRequestList.Items) > 0 {
+		var spaceRequest *codereadytoolchainv1alpha1.SpaceRequest
+		for i, s := range spaceRequestList.Items {
+			if s.Spec.TierName == environmentTierName {
+				spaceRequest = &spaceRequestList.Items[i]
+				break
+			}
+		}
+		return spaceRequest, nil
+	}
+
+	return nil, nil
+}
+
+// createSpaceRequestForDTC creates a new SpaceRequest for the given DTC. It sets the expected tierName and adds the
+// label indicating that it's linked with the DTC
+func createSpaceRequestForDTC(ctx context.Context, k8sClient client.Client, dtc *applicationv1alpha1.DeploymentTargetClaim) (*codereadytoolchainv1alpha1.SpaceRequest, error) {
+	newSpaceRequest := codereadytoolchainv1alpha1.SpaceRequest{
+		Spec: codereadytoolchainv1alpha1.SpaceRequestSpec{
+			TierName:           environmentTierName,
+			TargetClusterRoles: []string{},
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: dtc.Name,
+			Namespace:    dtc.Namespace,
+			Labels: map[string]string{
+				deploymentTargetClaimLabel: dtc.Name,
+			},
+		},
+	}
```

Comment:
Can we also add an ownerReference from the SpaceRequest, back to the parent `DeploymentTargetClaim`? This will help us ensure both are cleaned up.

(See the [Slack question](https://redhat-internal.slack.com/archives/C04SSH0S405/p1681339349974449) I posted on this. for some discussion)